### PR TITLE
Extended parameters (`-x` and `-E`) improvements

### PR DESCRIPTION
**Pull request by MCUdude**
Adresses #1697.

I've added the `-x autoreset` to the AVR109 programmer so that compatible bootloaders can utilize this feature. I've also added the `-x noautoreset` to the `arduino` and `wiring` programmers.

@stefanrueger I didn't touch the Urclock implementation, since the way it deals with extra parameters is so different than the other. It would be great if you could look into this!

This PR is still a draft, so feedback is welcome.

### CODE REVIEW

**Comment by stefanrueger** [comment]
@MCUdude Very cool. Will review/add/comment in the next few days

File: src/arduino.c
Comment (stefanrueger) [inline src/arduino.c:48]: `%2d` is a weird format, as it *only* reads 2 digits of the given number, but the user *doesn't know* and *it's hard to document*, so the user will be puzzled why `-xattempts=123` set the number of attempts to 12. Why not use `%i`? That allows hex numbers (`0x12`) or octal numbers (`077`), too. Normally one should also check that the number makes sense (must be positive here). Try `-xattempts=-1` and you'll see what I mean.

File: src/arduino.c
Comment (stefanrueger) [inline src/arduino.c:59]: Why not replace these 6 lines with one
```
  if(str_eq(extended_param, "no_autoreset")) {
```
Note the `str_eq()` instead of the `str_starts()`. Anything that starts with `no_autoreset` but isn't `no_autoreset` is caught later on anyway (invalid parameter).

File: src/arduino.c
Comment (stefanrueger) [inline src/arduino.c:66]: I'd write `-xattempts=<n> Specify the number <n> of connection retry attempts` for clarity. Also allows the option text to move left by two spaces

File: src/arduino.c
Comment (stefanrueger) [inline src/arduino.c:72]: In the last messaging clean-up the quotation marks tended to be removed, `"invalid extended parameter %s\n"` is the more recent AVRDUDE style.

File: src/arduino.c
Comment (stefanrueger) [inline src/arduino.c:157]: If you're changing spaces, why not move the curly bracket up, too? :smile:

File: src/arduino.c
Comment (stefanrueger) [inline src/arduino.c:173]: This looks wrong (because it probably is): *Must* be `PDATA(pgm)` not `PDATA(&pgm)`.I think this sets an unknown memory address to 1 (but not `autoreset`).

File: src/butterfly.c
Comment (stefanrueger) [inline src/butterfly.c:704]: see above

File: src/stk500.c
Comment (stefanrueger) [inline src/stk500.c:119]: I think at this place the check `str_eq(pgm->type, "Arduino")` must remain. It's OK to have the *additional* constraint `PDATA(pgm)->autoreset` that the user controls. Only when the `-c` programmer is of the Arduino type (ie, if it's `-c arduino`) should the driver trigger reset here. I believe, this has to do with the weird way optiboot has at some point decided when to jump to the application and when to jump to the bootloader. Normal STK500 programmers don't need this and it would prossibly be harmful to the re-sync attempts.

File: src/wiring.c
Comment (stefanrueger) [inline src/wiring.c:111]: See above

File: src/wiring.c
Comment (stefanrueger) [inline src/wiring.c:120]: I know `<arg>` below is in existing code. But it's always nicer to be as specific as possible. Arg can be anything, but `<n>` alludes to this being a number. So  `-xsnooze=<n> Wait snooze <n> ms` is defo more specific. Same with `delay` below.

File: src/wiring.c
Comment (stefanrueger) [inline src/wiring.c:185]: Cool!

**Comment by stefanrueger** [comment]
@MCUdude You'll have seen a couple of review comments. What I haven't done is compiling or testing the code (but you have the scope to see the number of auto-reset spikes, so that's best verified by yourself that the code does what you want it to do).

I'll push code for urclock's `-xnoautoreset` functionality. There is a ton of other `no...` extended options *without* the `_` between `no` and feature that is being switched off. So, for consistency in `urclock.c` I'll drop the `_`.

**Comment by stefanrueger** [comment]
And... this PR requires documentation in two places: `avrdude.1` and `avrdude.texi`

File: src/arduino.c
Comment (MCUdude) [reply src/arduino.c:48]: Sure! `%2d` is just copy-paste from stk500.c, but you have a point!

File: src/arduino.c
Comment (MCUdude) [reply src/arduino.c:173]: It works this way, and I got a segfault if not. I'll look into it again

File: src/stk500.c
Comment (MCUdude) [reply src/stk500.c:119]: Yes, but we only provide auto-reset functionality for the `-c arduino programmer`. So this should not run when using just another stk500v1 based programmer

**Comment by MCUdude** [comment]
Thanks for your feedback @stefanrueger! I have some comments though.

BTW one thing we should look into that would make the extended parameters more helpful to a struggling user would be to print the `-xhelp`output when the user provided an incorrect extended parameter. At the moment it's just an error, with no additional information. I think printing the help menu when there is an error could be helpful.

File: src/arduino.c
Comment (stefanrueger) [reply src/arduino.c:167]: I think I know what happens: `arduino_initpgm()` is called *before* `pgm->setup()` which allocates the pdata structure. So, `pgm->cookie` is NULL here (as it hasn't been initialises yet) and you get a seg-fault when you access it. `PDATA(&pgm)` is a memory location in the PROGRAMMER sructure which is at the offset of autoreset from where the cookie pointer lives. This is defo not what you want! *Either* set `PDATA(pgm)->autoreset` in `pgm->setup()` or later but before the parameters are parsed. *Or* invert the logic: Use a variable `noautoreset` that gets automatically initialised to zero. So you don't need to initialise that.

File: src/arduino.c
Comment (stefanrueger) [reply src/arduino.c]: Thanks. This means the `stk500.c` code also needs changing (as that too is wrong)

File: src/stk500.c
Comment (stefanrueger) [reply src/stk500.c:119]: That is a good point. But just imagine someone *adds* the autoreset functionality for *another* stk500.c based programmer (just like arduino did that). Then this code would run for that programmer, too. But I don't think this will be needed. I think it's *only* optiboot that needs that code at this point. `str_eq(pgm->type, "Arduino")` is kinda a proxy for optiboot (the best AVRDUDE can do outside what urclock has been doing by examining the bootloader itself and computing hash sums).

TLDR; I would use `if(str_eq(pgm->type, "Arduino") && PDATA(pgm)->autoreset && attempt > 0)` to document it's intended for `-c arduino` even though this might be redundant right now.

**Comment by stefanrueger** [comment]
> more helpful to a struggling user would be to print the `-xhelp` output 

Vvv good point! It's a lot of work to add this to what feels like 1500 programmers, but a great idea. (I think urclock already does this, so it's only needed for 1499 programmers :)

File: src/arduino.c
Comment (stefanrueger) [reply src/arduino.c]: > stk500.c code also needs changing

You already did! Apologies for stating the obvious.

File: src/stk500.c
Comment (MCUdude) [reply src/stk500.c:119]: Good point! Thanks for the explaination

Comment (stefanrueger) [reply src/stk500.c:119]: Actually, thinking about this more: as autoreset is introduced in the stk500.c PDATA and (one way or another) defaults to true this code *will* run for *all* stk500 programmers changing behaviour. It defo needs to be conditioned on Arduino.

**Comment by MCUdude** [comment]
> Vvv good point! It's a lot of work to add this to what feels like 1500 programmers, but a great idea. (I think urclock already does this, so it's only needed for 1499 programmers :)Vvv good point! It's a lot of work to add this to what feels like 1500 programmers, but a great idea. (I think urclock already does this, so it's only needed for 1499 programmers :)

How about something like this (separate print function):

```c
static void extended_param_helptext() {
  msg_error("%s -c %s extended options:\n", progname, pgmid);
  msg_error("  -xattempts=<n> Specify the number <n> of connection retry attempts\n");
  msg_error("  -xnoautoreset  Don't toggle RTS/DTR lines on port open to prevent a hardware reset\n");
  msg_error("  -xhelp         Show this help menu and exit\n");
}

static int arduino_parseextparms(const PROGRAMMER *pgm, const LISTID extparms) {
  const char *extended_param;
  int attempts;
  int rv = 0;

  for (LNODEID ln = lfirst(extparms); ln; ln = lnext(ln)) {
    extended_param = ldata(ln);

    if (sscanf(extended_param, "attempts=%i", &attempts) == 1) {
      PDATA(pgm)->retry_attempts = attempts;
      pmsg_info("setting number of retry attempts to %d\n", attempts);
      continue;
    }

    if(str_eq(extended_param, "noautoreset")) {
      PDATA(pgm)->autoreset = false;
      continue;
    }

    if (str_eq(extended_param, "help")) {
      extended_param_helptext();
      return LIBAVRDUDE_EXIT;
    }

    pmsg_error("invalid extended parameter %s\n", extended_param);
    extended_param_helptext();
    rv = -1;
  }

  return rv;
}
```

This will result in the following output:
```
$ ./avrdude -carduino -patmega328p -P /dev/cu.usbmodem14201 -xinvalid
avrdude error: invalid extended parameter invalid
avrdude -c arduino extended options:
  -xattempts=<n> Specify the number <n> of connection retry attempts
  -xnoautoreset  Don't toggle RTS/DTR lines on port open to prevent a hardware reset
  -xhelp         Show this help menu and exit
avrdude error: unable to parse extended parameter list
```

**Comment by stefanrueger** [comment]
> How about something like this

Output looks great; and yes, both ways would work.

**Comment by MCUdude** [comment]
@stefanrueger I've now pushed a fix for all programmers that supports extended parameters. I've done some minor code tweaks, but the main feature is that the help text gets printed if an invalid parameter is specified. Please review!

**Comment by MCUdude** [comment]
BTW I'll update the docs one #1842 is merged, to prevent merge conflicts.

**Comment by stefanrueger** [comment]
@MCUdude I think your technique of printing the help message on invalid parameters is very elegant. Great code! I particularly like  that the help message code resides *inside* the `for()` loop. This allows those valid `-x` parameters with invalid assignment values to  `break;` out of the look forgoing the full help message (which would distract from the nature of the invalid value).

Seeing that this is a really neat usability improvement for AVRDUDE, would you want to extend your technique to the parsing of exit parameters (`-E`)? It's three places
 - `par_parseexitspecs()`
 - `flip2_parseexitspecs()`
 - `linuxspi_parseexitspecs()`

> Please review!

Great code; a couple of comments
 - Please double check `jtagmkII_parseextparms()`: `help` never set; help preamble `msg_error("%s -c %s extended options:\n", progname, pgmid);` output in wrong place. I predict `avrdude -c jtag2 -xhelp` won't be what you want.
 - `WIRINGPDATA(&pgm)->autoreset = true;` is *still* a problem
 - I'd prefer `str_eq(pgm->type, "Arduino")` over `str_eq(pgmid, "arduino")` in `stk500_setup()` to be consistent when Arduino-like programmers are meant

> fix for all programmers that supports extended parameters

All? The commit overlooked
 - `serprog_parseextparms()`
 - `stk500v2_parseextparms()`
 - `usbasp_parseextparms()`

While reviewing the `xyz_parseextendedparms()` functions I also noticed that some of the error messages for invalid values on valid parameters did not make the context sufficiently clear (so a user who has carefully built a 2 mile long CLI line might struggle to figure out to which part of the command line the message actually refers to). I took the liberty to push a commit that makes those error messages clearer mostly by saying `Invalid value in -x <user setting>`. Also massaged the help messages a bit, eg, inserting a space between `-x` and the parameter. 

Please review this commit!

I also happened upon a bug in old code where `str_eq(str_eq(extended_param, "jtagchain=")` clearly needed to be `str_starts(extended_param, "jtagchain=")`. Fixed.

**Comment by MCUdude** [comment]
Thanks for the feedback @stefanrueger! I've done some work to `-E` as well, and I've added help text here as well (`-E help`). But it would be great if you had any ideas for better help text for these exitspec parameters. The ones I came up with are a bit too verbose I think.

File: src/serprog.c
Comment (stefanrueger) [inline src/serprog.c:484]: The next line needs to be `rv = LIBAVRDUDE_EXIT;` not `return ...`

I recommend to checking all 22 treated programmers by running
 ```
avrdude -c ... -x invalid
avrdude -c ... -x help
```
to see whether there are surprises.

File: src/dryrun.c
Comment (stefanrueger) [inline src/dryrun.c:1049]: BTW, really well spotted that a `continue` should not be `else`d. Neat!

File: src/jtag3.c
Comment (stefanrueger) [inline src/jtag3.c:1613]: Great catch!

**Review by stefanrueger — APPROVED** [review]

**Comment by MCUdude** [comment]
I've just tested all the programmers that support extended parameters. Things are looking good! Docs need to be updated, but I think it's best to wait until #1842 has been merged.

**Comment by MCUdude** [comment]
Speaking of docs, now is a good time to finally resolve #1711 by adding a new chapter or section to the avrdude.texi docs here?

image

**Pull request merged**
